### PR TITLE
fixed mobile input

### PR DESCRIPTION
**Pull request by thanhhovn2016**
fixed mobile input

### CODE REVIEW

**Comment by shawwwwwwww** [comment]
@thanhhovn2016 I could not get this to work on my machine. Does it work for you?
GameManagerSystem is missing, for one. When I remove that, ecsWorld is undefined. I think we are changing some stuff in some files that we don't need to, possibly causing a collision.

**Comment by thanhhovn2016** [comment]
Hi Shaw,I'm using old source code,I'll put it in latest source code on git.17.08.2021, 01:00, "shaw" ***@***.***>:
@thanhhovn2016 I could not get this to work on my machine. Does it work for you?
GameManagerSystem is missing, for one. When I remove that, ecsWorld is undefined. I think we are changing some stuff in some files that we don't need to, possibly causing a collision.

File: packages/client-core/src/common/components/TouchGamepad/index.tsx [hunk 1]
```diff
@@ -52,22 +59,32 @@ export const TouchGamepad: FunctionComponent<TouchGamepadProps> = () => {
...
+		}
+      }	  
       const event = new CustomEvent('touchstickmove', {
         detail: {
           stick: GamepadAxis.Left,
           value: { x: data.vector.y, y: -data.vector.x, angleRad: data.angle.radian }
         }
       })
       document.dispatchEvent(event)
+
```
Comment (shawwwwwwww) [inline packages/client-core/src/common/components/TouchGamepad/index.tsx:87]: Remove extra whitespace

File: packages/client-core/src/common/components/TouchGamepad/index.tsx [hunk 2]
```diff
@@ -52,22 +59,32 @@ export const TouchGamepad: FunctionComponent<TouchGamepadProps> = () => {
...
-    const targetElement = stickLeft[0].ui.el
-    targetElement.addEventListener('touchstart', (ev) => {
-      enableInput({ mouse: false })
-    })
-    targetElement.addEventListener('touchend', (ev) => {
-      enableInput({ mouse: true })
-    })
-
+ 
+	const canvasElement = (document.getElementById('engine-renderer-canvas') as HTMLCanvasElement)
```
Comment (shawwwwwwww) [inline packages/client-core/src/common/components/TouchGamepad/index.tsx:63]: EngineRenderer.instance.canvas, I think. Or something like that.

File: packages/client-core/src/common/components/TouchGamepad/index.tsx [hunk 3]
```diff
@@ -52,22 +59,32 @@ export const TouchGamepad: FunctionComponent<TouchGamepadProps> = () => {
...
     stickLeft.on('move', (e, data) => {
+	
+	  if (canvasElement.addEventListener) {  // all browsers except IE before version 9
```
Comment (shawwwwwwww) [inline packages/client-core/src/common/components/TouchGamepad/index.tsx:66]: Maybe this is right, can't check it, but this seems like every time the stick moves, we are adding a new even listener?

File: packages/client/src/components/World/index.tsx [hunk 4]
```diff
@@ -24,8 +24,19 @@ import GameServerWarnings from './GameServerWarnings'
 import { initEngine, retriveLocationByName, teleportToLocation } from './LocationLoadHelper'
 import { teleportPlayer } from '@xrengine/engine/src/avatar/functions/teleportPlayer'
 import { Network } from '@xrengine/engine/src/networking/classes/Network'
-import { NetworkSchema } from '@xrengine/engine/src/networking/interfaces/NetworkSchema'
+import { SystemUpdateType } from '../../../../engine/src/ecs/functions/SystemUpdateType'
+import { GameManagerSystem } from '../../../../engine/src/game/systems/GameManagerSystem'
```
Comment (shawwwwwwww) [inline packages/client/src/components/World/index.tsx:28]: This doesn't exist any more I don't think, or is throwing an error when I try to run this branch.

File: packages/client/src/components/World/index.tsx [hunk 5]
```diff
@@ -42,8 +53,15 @@ const getDefaulEngineInitializeOptions = (): InitializeOptions => {
     },
     physics: {
       simulationEnabled: false,
-      physxWorker: () => new Worker('/scripts/loadPhysXClassic.js')
-    }
+      physxWorker: new Worker('/scripts/loadPhysXClassic.js')
```
Comment (shawwwwwwww) [inline packages/client/src/components/World/index.tsx:56]: This is right? Does this need to be in this PR?

File: packages/client/src/components/World/index.tsx [hunk 6]
```diff
@@ -186,6 +204,9 @@ export const EnginePage = (props: Props) => {
   }, [props.locationState])
 
   useEffect(() => {
+	  //console.log('zzzzzzz')
+	  //props.showTouchpad = false
+	  console.log(props)
```
Comment (shawwwwwwww) [inline packages/client/src/components/World/index.tsx:209]: Remove

File: packages/client/src/components/World/index.tsx [hunk 7]
```diff
@@ -154,7 +172,7 @@ export const EnginePage = (props: Props) => {
 
   useEffect(() => {
     const currentLocation = props.locationState.get('currentLocation').get('location')
-
+	//console.log(currentLocation);
```
Comment (shawwwwwwww) [inline packages/client/src/components/World/index.tsx:175]: Remove

File: packages/engine/src/ecs/classes/World.ts [hunk 8]
```diff
@@ -19,16 +19,13 @@ export interface ECSWorld extends IWorld {
 let worldIds = 0
 export class World {
   static worlds: Map<number, World> = new Map<number, World>()
-  static defaultWorld: World
+  static defaultWorld: World = new World()
```
Comment (shawwwwwwww) [inline packages/engine/src/ecs/classes/World.ts:22]: Do we need to change the ECS?

File: packages/engine/src/ecs/classes/World.ts [hunk 9]
```diff
@@ -19,16 +19,13 @@ export interface ECSWorld extends IWorld {
...
   ecsWorld: ECSWorld
   entities: Entity[]
   portalEntities: Entity[]
   pipelines: EnginePipelines
 
   constructor() {
-    if (typeof World.defaultWorld === 'undefined') {
```
Comment (shawwwwwwww) [inline packages/engine/src/ecs/classes/World.ts:29]: This seems to be resolving a deeper ECS issue we have, if world isn't starting before we are calling this

File: packages/engine/src/initializeEngine.ts [hunk 10]
```diff
@@ -231,9 +231,6 @@ const registerServerSystems = (options: Required<InitializeOptions>) => {
 
 export const initializeEngine = async (initOptions: InitializeOptions = {}): Promise<void> => {
   const options: Required<InitializeOptions> = _.defaultsDeep({}, initOptions, DefaultInitializationOptions)
-
-  const world = new World()
```
Comment (shawwwwwwww) [inline packages/engine/src/initializeEngine.ts:235]: Was this changed in this PR intentionally?

File: packages/engine/src/initializeEngine.ts [hunk 11]
```diff
@@ -281,6 +278,7 @@ export const initializeEngine = async (initOptions: InitializeOptions = {}): Pro
     }
   }
 
+  const world = World.defaultWorld
```
Comment (shawwwwwwww) [inline packages/engine/src/initializeEngine.ts:281]: Was this changed in this PR intentionally?

File: packages/engine/src/input/schema/ClientInputSchema.ts [hunk 12]
```diff
@@ -63,13 +64,16 @@ export const handleTouchMove = (event: TouchEvent): void => {
     }
 
     prevTouchPosition = touchPosition
-
+	console.log('handleTouchMove =1')
```
Comment (shawwwwwwww) [inline packages/engine/src/input/schema/ClientInputSchema.ts:67]: Remove

File: packages/engine/src/input/schema/ClientInputSchema.ts [hunk 13]
```diff
@@ -63,13 +64,16 @@ export const handleTouchMove = (event: TouchEvent): void => {
...
     Engine.inputState.set(mappedMovementInput, {
       type: InputType.TWODIM,
       value: touchMovement,
       lifecycleState: Engine.inputState.has(mappedMovementInput) ? LifecycleValue.CHANGED : LifecycleValue.STARTED
     })
-  } else if (event.touches.length >= 2) {
```
Comment (shawwwwwwww) [inline packages/engine/src/input/schema/ClientInputSchema.ts:72]: Lots of deleted code here, was this from this PR?

**Review by shawwwwwwww — APPROVED** [review]
Doesn't work, so I can't test, and there's a lot of cruft that seems unrelated to this PR, but the touch input seems mostly right. Please clean it up, or make a new PR based on dev that just changes the input code.

**Comment by shawwwwwwww** [comment]
Great thanks

**Comment by thanhhovn2016** [comment]
I had commit to branch : input-mobileI have change engine\src\input\schema\ClientInputSchema.ts line 46:if (event.touches.length == 1) {  To if (event.touches.length >= 1) {If it problem with mobile input later then you change back to == 1 ,I didn't testing with == 1Please check.17.08.2021, 01:23, "shaw" ***@***.***>:
Great thanks

**Pull request closed without merging**